### PR TITLE
Zero Knowledge Proof

### PR DESCRIPTION
**Pull request by gbenattar**
See https://github.com/snipsco/rust-paillier/issues/29/

Test: cargo build & cargo test

### CODE REVIEW

File: src/zkproof/mod.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        for<'a,'b> &'a I: Rem<&'b I, Output=I>,
+        for<'a>    &'a I: Mul<I, Output=I>,
+{
+    fn generate_challenge(&self) -> (Vec<I>, I, Vec<I>, Vec<I>) {
+        let (mut y, mut challenge) : (Vec<I>, Vec<I>) = (Vec::new(), Vec::new());
+
+        let mut i : usize = 0;
+        while i < ZK_SECURITY_FACTOR {
+            let candidate = I::sample_below(&self.n);
+            if I::egcd(&self.n, &candidate).0 != I::one() { continue; }
```
Comment (mortendahl) [inline src/zkproof/mod.rs]: this check could be skipped without any practical impact as far as can I see (if gcd != 1 you managed to factor `n`). would also allow some simplification of the code.

File: src/zkproof/mod.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        I : Sub<I, Output=I>,
+        I : EGCD,
+        I : ToString,
+        I : FromString<I>,
+        for<'a>        I: Rem<&'a I, Output=I>,
+        for<'a,'b> &'a I: Rem<&'b I, Output=I>,
+        for<'a>    &'a I: Mul<I, Output=I>,
+{
+    fn generate_challenge(&self) -> (Vec<I>, I, Vec<I>, Vec<I>) {
+        let (mut y, mut challenge) : (Vec<I>, Vec<I>) = (Vec::new(), Vec::new());
```
Comment (mortendahl) [inline src/zkproof/mod.rs]: could take advantage of the fact that we know the final size of the vectors in advance (unless simplified as mentioned below)

**Comment by mortendahl** [comment]
I'd be interested in trying out [`ring`](https://github.com/briansmith/ring) instead of [`rust-crypto`](https://crates.io/crates/ring), the latter appearing somewhat deprecated (and even bloated perhaps). One concern is platform support but from a quick look `ring` at least supports mobile devices.

Any thoughts on this?

File: src/arithimpl/gmpimpl.rs [hunk 3]
```diff
@@ -45,6 +45,15 @@ impl ModPow for Mpz {
     }
 }
 
+impl ModMul for Mpz {
+    fn modmul(base: &Self, exponent: &Self, modulus: &Self) -> Self {
```
[the remark this answers is not in the document]
Comment (omershlo) [reply src/arithimpl/gmpimpl.rs]: @mortendahl :
This is supposed to be : (A * B) mod C = (A mod C * B mod C) mod C . 
It is used for example here:  M′=MQ+Q((Q−1modP)(MP−MQ)modP).
@gbenattar : base and exponent should be renamed.

File: src/zkproof/mod.rs (hunk 1 above)
Comment (omershlo) [reply src/zkproof/mod.rs]: I agree.

**Comment by gbenattar** [comment]
Hi,
Please find the new diff here: https://github.com/mortendahl/rust-paillier/pull/2/commits/e534cbe09974fa0bad4ee83da11b750f6f0f7cf4.

The new trait is (part of its own file):

```rust
pub struct CorrectInputProof<I> {
    pub e : I,
    pub z : Vec<I>,
}

pub struct CorrectKeyProof<I> {
    pub proof : I,
}

pub trait ProveCorrectKey<I, EK, DK> {
    fn generate_challenge(ek: &EK) -> (Vec<I>, CorrectInputProof<I>, Vec<I>);
    fn prove(dk: &DK, challenge: &Vec<I>, correct_input_proof: &CorrectInputProof<I>)
        -> Result<CorrectKeyProof<I>, ProofError>;
    fn verify(correct_key_proof: &CorrectKeyProof<I>, y: &Vec<I>) -> Result<(), ProofError>;
}
```

**Comment by gbenattar** [comment]
Changing base branch to dev as requested by @mortendahl.

**Comment by gbenattar** [comment]
In the last commit, I switched from `rust-crypto` to `ring` as suggested in @mortendahl comments.
I have no concerns for this change.

**Pull request merged**